vms/platformvm/validators/manager.go

```diff
@@ -198,146 +225,162 @@ func (m *manager) getValidatorSetFrom(
 	// cache the validator set
 	validatorSetsCache.Put(targetHeight, validatorSet)
 
-	endTime := m.clk.Time()
+	duration := m.clk.Time().Sub(startTime)
 	m.metrics.IncValidatorSetsCreated()
-	m.metrics.AddValidatorSetsDuration(endTime.Sub(startTime))
+	m.metrics.AddValidatorSetsDuration(duration)
 	m.metrics.AddValidatorSetsHeightDiff(currentHeight - targetHeight)
 	return validatorSet, nil
 }
 
+func (m *manager) getValidatorSetCache(subnetID ids.ID) cache.Cacher[uint64, map[ids.NodeID]*validators.GetValidatorOutput] {
+	// Only cache tracked subnets
+	if subnetID != constants.PrimaryNetworkID && !m.cfg.TrackedSubnets.Contains(subnetID) {
+		return &cache.Empty[uint64, map[ids.NodeID]*validators.GetValidatorOutput]{}
+	}
+
+	validatorSetsCache, exists := m.caches[subnetID]
+	if exists {
+		return validatorSetsCache
+	}
+
+	validatorSetsCache = &cache.LRU[uint64, map[ids.NodeID]*validators.GetValidatorOutput]{
+		Size: validatorSetsCacheSize,
+	}
+	m.caches[subnetID] = validatorSetsCache
+	return validatorSetsCache
+}
+
 func (m *manager) makePrimaryNetworkValidatorSet(
-	currentHeight uint64,
+	ctx context.Context,
 	targetHeight uint64,
-) (map[ids.NodeID]*validators.GetValidatorOutput, error) {
+) (map[ids.NodeID]*validators.GetValidatorOutput, uint64, error) {
+	validatorSet, currentHeight, err := m.getCurrentPrimaryValidatorSet(ctx)
+	if err != nil {
+		return nil, 0, err
+	}
+	if currentHeight < targetHeight {
+		return nil, 0, database.ErrNotFound
+	}
+
+	// Rebuild primary network validators at [targetHeight]
+	//
+	// Note: Height h contains the diffs from height h-1 to h. So, to rebuild
+	// validators state at height h, we apply diffs till height h+1.
+	lastDiffHeight := targetHeight + 1
+	err = m.state.ApplyValidatorWeightDiffs(
+		ctx,
+		validatorSet,
+		currentHeight,
+		lastDiffHeight,
+		constants.PlatformChainID,
+	)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	err = m.state.ApplyValidatorPublicKeyDiffs(
+		ctx,
+		validatorSet,
+		currentHeight,
+		lastDiffHeight,
+	)
+	return validatorSet, currentHeight, err
+}
+
+func (m *manager) getCurrentPrimaryValidatorSet(
```

Comment:
This is separated into its own function because this function will need a P-chain lock held to get a consistent view of the validator set + height.